### PR TITLE
nemu: enable compile nemu for arm64 on x86 host

### PR DESCRIPTION
**Pull request by jongwu**
This patch enable cross-compile nemu for arm64 in container
on x86 host.

fixed: #577
Signed-off-by: Jianyong Wu  <jianyong.wu@arm.com>
@jodh-intel @bergwolf @ganeshmaharaj @grahamwhaley @Pennyzct

### CODE REVIEW

File: static-build/nemu/Dockerfile_aarch64 [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+RUN git clone https://github.com/qemu/capstone.git capstone
+RUN git clone https://github.com/qemu/keycodemapdb.git ui/keycodemapdb
+ADD configure-hypervisor.sh /root/configure-hypervisor.sh
+RUN wget http://archive.ubuntu.com/ubuntu/pool/universe/g/gcc-defaults/pkg-config-aarch64-linux-gnu_7.4.0-1ubuntu2.3_amd64.deb && dpkg -i pkg*
+
+RUN cp -r /usr/lib/aarch64-linux-gnu/pkgconfig/ /usr/aarch64-linux-gnu/lib/
+
+RUN PREFIX="${PREFIX}" /root/configure-hypervisor.sh -s kata-nemu aarch64 | xargs ./configure \
+       --with-pkgversion=kata-static
+RUN make -j$(nproc) CONFIG_HID_DEV=y
```
Comment (jodh-intel) [inline static-build/nemu/Dockerfile_aarch64]: If this is only going to run under the CI, this is fine.

However, if this is going to run on a developers system, this is going to DoS it for the duration of the `make` run.

In the latter case, we tend to do:

```sh
make -j$(($(nproc)-1)) ...
```

Comment (jongwu) [reply static-build/nemu/Dockerfile_aarch64]: very appreciate @jodh-intel . I have refined this patch base on most of your precious feedback.  Further more I try to make this cross-compile can work on other arch as your opinion says, but only test on arm64.

File: scripts/configure-hypervisor.sh [hunk 2]
```diff
@@ -192,6 +195,17 @@ show_array() {
 
 generate_qemu_options() {
 	#---------------------------------------------------------------------
+	#cross-compile
+	host=$(uname -m)
```
[the remark this answers is not in the document]
Comment (jongwu) [reply scripts/configure-hypervisor.sh]: thanks @jodh-intel , I want to test if destination arch identical to the host here. Is there a better way to do that?

File: static-build/nemu/build-static-nemu.sh [hunk 3]
```diff
@@ -15,6 +15,31 @@ source "${script_dir}/../../scripts/lib.sh"
 config_dir="${script_dir}/../../scripts/"
 nemu_tar="kata-nemu-static.tar.gz"
 
+if [ $# -ne 0 ];then
+       arch="$1"
+       echo $arch
+       if [ "$arch" = "aarch64" ];then
+	       echo aaaaaaaaaaaa
```
Comment (jodh-intel) [inline static-build/nemu/build-static-nemu.sh]: Debug output? Same comment for the "`b`", "`c`" and "`1`" `echo`'s below :smile:

File: static-build/nemu/Dockerfile_cross [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+ARG PREFIX
+ARG GEN_ARCH
+ARG DPKG_ARCH
+
+WORKDIR /root/nemu
+
+RUN echo "deb http://deb.debian.org/debian stable main" > /etc/apt/sources.list
+RUN echo "deb http://deb.debian.org/debian stable-updates main" >> /etc/apt/sources.list
+RUN echo "deb http://security.debian.org stable/updates main" >> /etc/apt/sources.list
+RUN dpkg --add-architecture $DPKG_ARCH
```
Comment (jodh-intel) [inline static-build/nemu/Dockerfile_cross]: Nice :smile:

File: static-build/nemu/Dockerfile_cross [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+RUN apt update && apt install -y \
+	gcc-$GEN_ARCH-linux-gnu \
+	autoconf \
```
Comment (jodh-intel) [inline static-build/nemu/Dockerfile_cross]: Nit: Bonus points if you sort the list alphabetically ;)

**Comment by jongwu** [comment]
@jodh-intel thanks

**Comment by jodh-intel** [comment]
@kata-containers/packaging - please review! :smile:

File: scripts/configure-hypervisor.sh [hunk 6]
```diff
@@ -16,7 +16,9 @@
 
 script_name=${0##*/}
 
-arch=$(uname -m)
+arch="$3"
+arch="${arch:-$(uname -m)}"
```
Comment (jcvenegas) [inline scripts/configure-hypervisor.sh]: nit: arch="${3:-$(uname -m)}"

File: scripts/configure-hypervisor.sh [hunk 7]
```diff
@@ -192,6 +194,17 @@ show_array() {
 
 generate_qemu_options() {
 	#---------------------------------------------------------------------
+	#check if cross-compile is needed
+	host=$(uname -m)
+	if [ $arch != $host ];then
+		case $arch in
+			aarch64) qemu_options+=(size:--cross-prefix=aarch64-linux-gnu-);;
+			ppc64le);;
```
Comment (jcvenegas) [inline scripts/configure-hypervisor.sh]: die "Not not implemented"  ?

Comment (jongwu) [reply scripts/configure-hypervisor.sh]: @jcvenegas I'm not sure the prefix for ppc64le and s390, do you have any idea?

File: scripts/configure-hypervisor.sh (hunk 6 above)
Comment (jongwu) [reply scripts/configure-hypervisor.sh]: ok @jcvenegas

File: scripts/configure-hypervisor.sh (hunk 7 above)
Comment (jodh-intel) [reply scripts/configure-hypervisor.sh]: @nitkon and @alicefr - could you help here?

Comment (jcvenegas) [reply scripts/configure-hypervisor.sh]: @jongwu  no idea :/ but just to make explicit that is not implemented you can add a way to exit so in the future could be added

**Comment by jodh-intel** [comment]
Any more review comments folks?

**Comment by jongwu** [comment]
I am not sure if this test failure caused by my commit. any hint? @jcvenegas

**Comment by jongwu** [comment]
@jodh-intel can you help me check the error log? I could not associate the error with my commit.

**Comment by jodh-intel** [comment]
Hi @jongwu - the snap error is currently a known issue and unrelated to your PR.

@jcvenegas (or anyone else) - how do we trigger the OBS CI?

**Comment by jongwu** [comment]
so what we should do with that snap test fail issue ? @jodh-intel @jcvenegas

**Comment by grahamwhaley** [comment]
@jongwu - first, we follow the link to the build failure, and grab the error from the logs ;-)

```
patching file virt/kvm/arm/vgic/vgic-mmio-v3.c
Failed to run 'override-build': Exit code was 1.
Makefile:47: recipe for target 'snap' failed
make: *** [snap] Error 2
Build step 'Execute shell' marked build as failure
```

Hmm, failed to patch or build - @devimc - seen this before?

**Comment by Pennyzct** [comment]
Hi~ @grahamwhaley we have discussed this issue here, - [packaging/#592](https://github.com/kata-containers/packaging/pull/592#issuecomment-504854782). I saw that we have bumped the version to 1.8.0-rc0. maybe we can update the version in [snapcraft.yaml](https://github.com/kata-containers/packaging/blob/master/snap/snapcraft.yaml#L2) to 1.8.0-rc0 to get the snap CI green again. ;) 😊

**Comment by jongwu** [comment]
the snap issue may have been repaired, help retrigger snap test @jcvenegas  @jodh-intel

**Comment by grahamwhaley** [comment]
/test
old build has expired - I'm not sure we have an individual snap [trigger phrase](https://github.com/kata-containers/community/wiki/Controlling-the-CI#specific-job-triggers) (@GabyCT - if we have more phrases now than on the wiki page, please update ;-) ).... so, I'll trigger a full CI run.

**Comment by jongwu** [comment]
what's next? @jodh-intel @grahamwhaley

File: static-build/nemu/Dockerfile_cross [hunk 8]
```diff
@@ -0,0 +1,65 @@
+FROM debian
```
Comment (grahamwhaley) [inline static-build/nemu/Dockerfile_cross]: You should probably have a copyright header in here :-)

Comment (jongwu) [reply static-build/nemu/Dockerfile_cross]: @grahamwhaley ok, thanks.

**Comment by jongwu** [comment]
" Failed to connect to 10.96.18.55 port 8080: Connection timed out" @caoruidong , help retrigger this test.

**Comment by jongwu** [comment]
this failure maybe no related to nemu install, @caoruidong help retest, thanks

**Comment by egernst** [comment]
Sorry this took (or, is taking) so long @jongwu .  Thanks for your patience.  @devimc @Pennyzct know anything re: the snap failure?

**Comment by devimc** [comment]
@egernst crio is failing, restarting the CI, may be it was a random failure

```
1..45
not ok 1 ctr not found correct error message
# (from function `setup_crio' in file helpers.bash, line 249,
#  from function `start_crio' in file helpers.bash, line 308,
#  in test file ctr.bats, line 10)
#   `start_crio' failed
```

**Comment by jongwu** [comment]
thanks @devimc @egernst , all green now! Help review @devimc ,thanks!

**Comment by devimc** [comment]
@jongwu  thanks

**Pull request merged**